### PR TITLE
[VarExporter] add Instantiator::instantiate() to create+populate objects without calling their constructor nor any other methods

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

A blend of features also provided by https://github.com/doctrine/instantiator and https://github.com/Ocramius/GeneratedHydrator in one simple method. Because it's just a few more lines on top of the existing code infrastructure in the component :)

For example, from the docblock:

```php
// creates an empty instance of Foo
Instantiator::instantiate(Foo::class);

// creates a Foo instance and sets one of its public, protected or private properties
Instantiator::instantiate(Foo::class, ['propertyName' => $propertyValue]);

// creates a Foo instance and sets a private property defined on its parent Bar class
Instantiator::instantiate(Foo::class, [], [
    Bar::class => ['privateBarProperty' => $propertyValue],
]);
```

Instances of ArrayObject, ArrayIterator and SplObjectHash can be created
by using the special `"\0"` property name to define their internal value:

```php
// creates an SplObjectHash where $info1 is attached to $obj1, etc.
Instantiator::instantiate(SplObjectStorage::class, ["\0" => [$obj1, $info1, $obj2, $info2...]]);

// creates an ArrayObject populated with $inputArray
Instantiator::instantiate(ArrayObject::class, ["\0" => [$inputArray, $optionalFlag]]);
```

Misses some tests for now, but reuses the existing code infrastructure used to "unserialize" objects.

### CODE REVIEW

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @param array $propertiesByClass The properties to set on the object, keyed by the class to bind to
+     *                                 when setting them. The first key of the array defines the class
+     *                                 of the returned instance.
+     *
+     * @return object An instance of the class defined by the first key of $propertiesByClass
+     *
+     * @throws \ReflectionException When the instance cannot be created
+     */
+    public static function instantiate(array $propertiesByClass)
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: ` : object`

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\VarExporter;
+
+use Symfony\Component\VarExporter\Internal\Hydrator;
+use Symfony\Component\VarExporter\Internal\Registry;
+
+/**
+ * A utility class to create objects without calling their constructor.
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Most likely a good idea to keep this `@internal` until proven to be stable for usage within the component

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+{
+    /**
+     * Creates an object and sets its properties without calling its constructor nor any other methods.
+     *
+     * For example:
+     *
+     *     // creates an empty instance of Foo
+     *     Instantiator::instantiate([Foo::class => []]);
+     *
+     *     // creates a Foo instance and sets one of its public, protected or private properties
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: The `public`/`private`/`protected` case should likely be always distinguished, because of inheritance and possible property overrides.

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+     *     Instantiator::instantiate([Foo::class => ['propertyName' => $propertyValue]]);
+     *
+     *     // creates a Foo instance and sets a private property defined on its parent Bar class
+     *     Instantiator::instantiate([
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: This syntax is very confusing: the object type should be the first `string` parameter. What's the reasoning behind putting the first type inside the array?

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+     *         Foo::class => [],
+     *         Bar::class => ['privatePropertyName' => $propertyValue],
+     *     ]);
+     *
+     * Instances of ArrayObject, ArrayIterator and SplObjectHash can be created
+     * by using the special "\0" property name to define their internal value:
+     *
+     *     // creates an SplObjectHash where $info1 is attached to $obj1, etc.
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Strongly suggest keeping anything coming from php core or extensions out of the scope of the component, as well as its documentation: there's an infinity of wrongness in classed that are not defined in PHP userland, and we really should rather tell users to not rely on them.

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+     *     Instantiator::instantiate([SplObjectStorage::class => ["\0" => [$obj1, $info1, $obj2, $info2...]]]);
+     *
+     *     // creates an ArrayObject populated with $inputArray
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Same as above: internal classes should be kept out, or in a dark corner with large warning signs that stop users from getting there in first place

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+     *     Instantiator::instantiate([ArrayObject::class => ["\0" => $inputArray]]);
+     *
+     * @param array $propertiesByClass The properties to set on the object, keyed by the class to bind to
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Please document the type for keys and values here: should probably be `array<string, array<string, mixed>>` from the examples above, but would suggest extracting the first key

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+     *                                 when setting them. The first key of the array defines the class
+     *                                 of the returned instance.
+     *
+     * @return object An instance of the class defined by the first key of $propertiesByClass
+     *
+     * @throws \ReflectionException When the instance cannot be created
+     */
+    public static function instantiate(array $propertiesByClass)
+    {
+        if (!$propertiesByClass) {
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: This is easily prevented by enforcing a `string` type for an additional first mandatory argument, which makes the API friendlier to the used

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+        foreach ($propertiesByClass as $class => $properties) {
+            break;
+        }
+
+        try {
+            $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+        } catch (\Exception $e) {
+            if (0 === strpos($e->getMessage(), 'Serialization of ')) {
+                $e = new \ReflectionException(substr_replace($e->getMessage(), 'Instanti', 0, 8));
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Should still add the previous exception

Comment (Ocramius) [reply src/Symfony/Component/VarExporter/Instantiator.php]: Please declare and throw a component-specific exception (which may extend from `ReflectionException`)

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            throw new \ReflectionException('No class provided.');
+        }
+
+        foreach ($propertiesByClass as $class => $properties) {
+            break;
+        }
+
+        try {
+            $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+        } catch (\Exception $e) {
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Catching `\Exception` here is a problem: this indicates that the registry is not throwing specific-enough exceptions, and using the `strpos()` approach is absolutely not OK

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @throws \ReflectionException When the instance cannot be created
+     */
+    public static function instantiate(array $propertiesByClass)
+    {
+        if (!$propertiesByClass) {
+            throw new \ReflectionException('No class provided.');
+        }
+
+        foreach ($propertiesByClass as $class => $properties) {
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: This is extremely confusing to read, and derives from the confusing method signature that I noted above: requires refactoring

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+        } catch (\Exception $e) {
+            if (0 === strpos($e->getMessage(), 'Serialization of ')) {
+                $e = new \ReflectionException(substr_replace($e->getMessage(), 'Instanti', 0, 8));
+            }
+
+            throw $e;
+        }
+
+        if (Registry::$cloneable[$class]) {
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Why not use `doctrine/instantiator` directly here? Would save a lot of headaches and duplicate efforts, and already performs these checks at the cost of one method call. Are there test cases that we need to port over there too?

File: src/Symfony/Component/VarExporter/README.md [hunk 13]
```diff
@@ -5,6 +5,9 @@ The VarExporter component allows exporting any serializable PHP data structure t
 plain PHP code. While doing so, it preserves all the semantics associated with
 the serialization mechanism of PHP (`__wakeup`, `__sleep`, `Serializable`).
 
+It also provides an instantiator that allows creating and populating objects
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/README.md]: Better keep it well hidden for now

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            $instance = $reflector->newInstanceWithoutConstructor();
+        } elseif (null === Registry::$prototypes[$class]) {
+            throw new \ReflectionException(sprintf("Instantiation of '%s' is not allowed", $class));
+        } elseif ($reflector->implementsInterface('Serializable')) {
+            $instance = unserialize('C:'.\strlen($class).':"'.$class.'":0:{}');
+        } else {
+            $instance = unserialize('O:'.\strlen($class).':"'.$class.'":0:{}');
+        }
+
+        $instance = array($instance);
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Few problems here:

 1. The variable type changes from `object` to `array<object>`, which is confusing: declare a new variable instead
 2. The fact that the hydration operates on `array<object>`, and that the wrapping is necessary in first place, is confusing

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+        foreach ($propertiesByClass as $class => $properties) {
+            if (!$properties) {
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: Would probably rather throw an exception: why is a key even declared if nothing is to be hydrated for this inheritance level?

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+                continue;
+            }
+            foreach ($properties as $name => $value) {
+                $properties[$name] = array($value);
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: This recursive wrapping is confusing too: why is everything operating with arrays and pseudo-signatures?

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+            }
+            (Hydrator::$hydrators[$class] ?? Hydrator::getHydrator($class))($properties, $instance);
```
Comment (Ocramius) [inline src/Symfony/Component/VarExporter/Instantiator.php]: I wrote a little component called `ocramius/lazy-map` for these cases: unsure if worth using its approach here, as it may both speed up or slow down everything (needs benchmarks), but at least it gets rid of the null coalesce everywhere

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by javiereguiluz** [comment]
Thanks for the detailed description! Looking at the end-user API, I have two comments/questions:

1) Do we really need to support an array as the argument of `instantiate()`? Is that common to instantiate more than one object at a time?

```php
// Before
Instantiator::instantiate([
    Foo::class => [],
    Bar::class => ['propertyName' => $propertyValue],
]);

// After
Instantiator::instantiate(Foo::class, []);
Instantiator::instantiate(Bar::class, ['propertyName' => $propertyValue]);
```

2) The use of `\0` for some special objects looks like an internal detail which compliates the learning curve when exposing it publicly. Why not "hardcode" this detail internally and let the end-user not know about this?

```php
// Before
Instantiator::instantiate([SplObjectStorage::class => ["\0" => [$obj1, $info1, $obj2, $info2]]]);
Instantiator::instantiate([ArrayObject::class => ["\0" => $inputArray]]);

// After
Instantiator::instantiate(SplObjectStorage::class, [$obj1 => $info1, $obj2 => $info2]);
Instantiator::instantiate(ArrayObject::class, $inputArray);
```

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: hello PHP 7.1 here ;)

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 9 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: Addressed in #28422

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: resolved by using a single $wrappedInstance variable

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 17 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: Thanks for the link, interesting. I this specific case, we should stay with `??` as it is really the best way when using arrays directly (no deps+perf-wise).

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 16 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: comment added to hint about the answer: because they're also used for "unserialization", hydrators deal with array of instances, so we need to wrap values.

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: The proposed method borrows from the existing infrastructure in the component: if there is any issues with `Instantiator::instantiate()`, it's also an issue for `VarExporter::export()`. It would make no sense to mark this method as internal, and not the other.

And if we mark the other internal, then the component would have zero public interfaces. It wouldn't make sense :)

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: Same as above: this is already dealt with for `VarExporter::export()`, so any issue found there should be spotted and fixed. There are already tests for these btw (which doesn't mean it's yet bulletproof of course.)

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: What's the syntax to document this, one that would be supported by phpstorm and phpdocumentor? For now, I described the expected structure in the param description and in the example. That could be enough in practice (`array<string, array<string, mixed>>` is a bit too abstract to be useful I fear)

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 12 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: The headaches are already past stories :)
The `VarExporter::export()` need access to internal details when generating its output, so this is already solved concern.

File: src/Symfony/Component/VarExporter/Instantiator.php [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+     *
+     * @return object The created instance
+     *
+     * @throws ExceptionInterface When the instance cannot be created
+     */
+    public static function instantiate(string $class, array $properties = array(), array $privateProperties = array())
+    {
+        $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+
+        if (Registry::$cloneable[$class]) {
```
Comment (stof) [inline src/Symfony/Component/VarExporter/Instantiator.php]: what if keys are not set for all these array accesses ?

Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: This cannot happen, because of the line just above that ensures it.
Yes, that's not "proved" by the interfaces, that's why everything is internal btw.

File: src/Symfony/Component/VarExporter/Instantiator.php (hunk 15 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/VarExporter/Instantiator.php]: There would be no mistake so an exception would be too strict here. Think of generated values: forcing a non-empty array here would increase the burden on the caller to prepare its value, for no logical (no "mathematical") reason. An empty array is valid input to me.

**Comment by nicolas-grekas** [comment]
@Ocramius @javiereguiluz thanks for the review+comments, I addressed some inline, let's address the remaining ones:

I added a 2nd commit to change the signature to `instantiate(string $class, array $properties, array $privateProperties);`. See updated examples in the PR. If you think it's better for DX, fine to me.

> The use of \0 for some special objects looks like an internal detail

It's not actually: e.g. you can also add dynamic properties to ArrayObject instances, that are *not* stored in the internal array structure. The instantiate method has to provide a way to set both these properties, and the internal array. `\0` is specifically chosen because no dynamic properties can start with the nul character, so it is safe to use.

This doesn't mean we cannot do what you propose @javiereguiluz, but we still need a way to set dynamic properties when needed.

Actually, there is a trivial way that works already: using the special `*` key in the `$privateProperties`: this populates public (thus dynamic) properties already.

So, let's say we want to create something like using instantiate:
```php
$a = new ArrayObject($input);
$a->foo = 123;
```

The current code allows doing so using:
```php
Instantiator::instantiate(ArrayObject::class, ['foo' => 123, "\0" => $input]);
```

And we *could* make it do the same using:
```php
Instantiator::instantiate(ArrayObject::class, $input, ['*' => ['foo' => 123]]);
```

Please advise.

**Comment by javiereguiluz** [comment]
Thanks for the detailed explanation. I hadn't thought about this use case:

```php
$a = new ArrayObject($input);
$a->foo = 123;
```

I thought this utility was only for instantiating ... but that code instantiates and then initializes, so it's doing two different things :)

What if we use this signature?

```
public static function instantiate(
    string $classFqcn,
    array $constructorArguments = [],
    array $propertyValues = []
)
```

Then, this example would look as:

```php
Instantiator::instantiate(ArrayObject::class, [$input], ['foo' => 123])
```

But most of the times you'd simply use something like this:

```php
Instantiator::instantiate(ArrayObject::class, [$input])
```

**Comment by nicolas-grekas** [comment]
@javiereguiluz wouldn't make sense: the specific property of instantiators is to *not* call the constructor nor any other methods...

**Comment by javiereguiluz** [comment]
I just followed your last example: why pass `$input` if it's never used?

**Comment by javiereguiluz** [comment]
After talking with Nico on Slack about this ... it's clear to me that I don't fully understand this 😅 So let's forget about my previous examples and proposals. Sorry!

**Comment by nicolas-grekas** [comment]
Good news: shaking the code always helps to spot new edge cases.
PR is ready, with test cases and fixes found meanwhile.

**Comment by nicolas-grekas** [comment]
Fixes split in PR #28437. This PR is now rebased on top of it.

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/VarExporter/Tests/InstantiatorTest.php [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace Symfony\Component\VarExporter\Tests;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\VarDumper\Test\VarDumperTestTrait;
+use Symfony\Component\VarExporter\Instantiator;
+
+class InstantiatorTest extends TestCase
+{
+    use VarDumperTestTrait;
```
Comment (ro0NL) [inline src/Symfony/Component/VarExporter/Tests/InstantiatorTest.php]: not sure im missing it.. but this can be removed right?

**Comment by fabpot** [comment]
@nicolas-grekas We needs good docs on that one (and the whole component). Can you create a PR for that in the docs repo? I think it won't take too much of your time as you're the best to write something smart about this :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
